### ISSUE TITLE
Apply `ruff` formatting to `markdown` code blocks

### ISSUE THREAD

**Issue by paddyroddy** [astral-sh/ruff#3792]
It would be cool to have the functionality to run `ruff` over `python` codeblocks like this does for `black` https://github.com/adamchainz/blacken-docs

**Comment by JonathanPlasse** [comment]
https://github.com/pydantic/pytest-examples
Does something similar.

**Comment by evanrittenhouse** [comment]
I'd like to work on this!

**Comment by evanrittenhouse** [comment]
I'm interested in trying to implement this. Looking at the projects listed above, it seems like code blocks are determined via Regex. Do we want to copy that approach or use a more robust Markdown parser ([pulldown-cmark](https://github.com/ralphlinus/pulldown-cmark) seems like a popular one)?

**Comment by JonathanPlasse** [comment]
I would vote to use a markdown parser for robustness as you suggested.

**Comment by MichaReiser** [comment]
Using a markdown parser makes sense from my view. It may be worth to also consider [tree-sitter](https://tree-sitter.github.io/tree-sitter/) because it supports many languages. 

The part that's unclear to me how we want to solve it in the short term is how we want to map column and line numbers in `Message`s. Ruff supports Jupyter-notebooks today but the line-mapping is somewhat "hacked-in" to make it work and I'm a bit reluctant to add more exceptions to the line-number mapping. 

https://github.com/charliermarsh/ruff/blob/483f4799995d1aea7987e6f7ee403f0411060023/crates/ruff/src/message/text.rs#L70-L87

I assume that a similar mapping will be necessary for Markdown files because we only pass the python code to Ruff, but we should show users the absolute line number from the start of the Markdown document. 

## Long-term

Long-term, we'll need to support the following two features

### Multi-language support
Ruff should support linting different file types. For example, Ruff should be able to lint SQL, python, and Markdown files. The filetype decides which specific linter Ruff uses to lint the file. This includes that the LSP is in sync with the file-types supported by the CLI and what extensions map to which languages. Rome supports [this today](https://github.com/rome/tools/blob/0ce919840c8ebaa02c3329b38fd8864c05b38d6e/crates/rome_service/src/file_handlers/mod.rs#L237-L261).

### Embedded Languages
The idea is that a document written in one language can contain code written in another language. Examples are:

* Code blocks in markdown documents
* SQL in Python code
* Python code in Jupyter Notebooks 

The markdown file handler would recognize code blocks and delegate to Ruff to decide how to parse, lint, and format the code block's content.

**Comment by evanrittenhouse** [comment]
Yup, agree with all of your points @MichaReiser. As part of the design phase I've been thinking about how to generalize to different filetypes. It's an interesting problem - I will take a look at Treesitter parsers as well. Going down that route, we could support different languages relatively easily. I am going to submit PRs for this incrementally so that we can iterate over the design. The first one will be a parser that pulls code blocks out of Markdown - I'm leaning towards using Treesitter since that architecture could make it easier to provide "plugins" for other languages in the future.

**Comment by MichaReiser** [comment]
> As far as the line mappings, I was planning on making a generic LineMap struct (or something along those lines) which would contain line numbers mapped to their offset in the containing document.

Mapping line numbers should suffice for markdown documents but wouldn't be sufficient for e.g. SQL in python. I think it may actually be sufficient to simply add the byte offset of the markdown block to the range of every message/diagnostic.

**Comment by evanrittenhouse** [comment]
Yup, that's basically what I'm doing - adding a `document_offset` field to Messages and Diagnostics. 

As far as the SQL in Python example, I was thinking the approach could work because the `document_offset` would be where the SQL code starts relative to the Python file. Line 1 of the SQL code could be on line 50 of the overarching .py file, so the `document_offset` for all diagnostics/messages in that block would be 50. Do you think there are flaws in that approach?

**Comment by evanrittenhouse** [comment]
Going with a treesitter integration requires downloading the treesitter grammar for the language in question. While we could gate this behind a workspace `feature` flag, I'd like some other folk's thoughts on it. I don't want to needlessly increase the binary size.

Advantage of `pulldown-cmark` as mentioned above is it's quite fast and won't affect the binary size. Disadvantage is that it obviously can't be used for anything except Markdown.

@MichaReiser @JonathanPlasse do you guys have any thoughts?

**Comment by JonathanPlasse** [comment]
It seems easy at first to handle only markdown and in a second time expand with the tree-sitter grammar.

**Comment by MichaReiser** [comment]
> Yup, that's basically what I'm doing - adding a document_offset field to Messages and Diagnostics.

Which use cases require a diagnostic or message to know its document offset? Would it be sufficient to mutate the `diagnostic.range` or `message.range` directly by adding the offset? I'm asking because increasing the size of `Message` and `Diagnostic` decrease performance because:
* More data that must be written to and read from the cache
* Overall more data to write when creating/copying diagnostics and read when generating the output (even for diagnostics not using embedded languages) 

> Advantage of pulldown-cmark as mentioned above is it's quite fast and won't affect the binary size. 

My expectation is that pulling in `pulldown-cmark` increases the binary size because we include a new crate in our binary (that is 116 kB in size), or is pulldown-cmark already a (runtime) dependency ? 

> Going with a treesitter integration requires [installing](https://github.com/tree-sitter/tree-sitter/tree/master/lib/binding_rust?rgh-link-date=2023-04-29T22%3A29%3A32Z) the treesitter grammar for the language in question.

Yeah, using treesitter is more complicated because it requires a custom build step. I'm not too opinionated on if we should use treesitter or not. I think we can start prototyping with either and defer the decision to later.

**Comment by evanrittenhouse** [comment]
>Would it be sufficient to mutate the diagnostic.range or message.range directly by adding the offset?
Yep! I was working on it yesterday and that's also the route I've chosen to go down. Sorry for the confusion, I should've waiting until I had a fully fleshed out design before weighing in.

>My expectation is that pulling in pulldown-cmark increases the binary size because we include a new crate in our binary (that is 116 kB in size), or is pulldown-cmark already a (runtime) dependency ?
It's not a runtime dependency, so it'll increase the crate because of its size, yes. I meant that it'd be less than the size delta from installing treesitter and associated grammar(s) which aren't totally necessary at this stage.

>I think we can start prototyping with either and defer the decision to later.
Cool!

**Comment by evanrittenhouse** [comment]
Listing down some of the issues I see:
1. Markdown, unlike Jupyter, is composed of discrete code blocks that don't necessarily affect each other. **If we want to treat Markdown code blocks as discrete units**, we'll have to rewrite most of the hot path since it's tightly coupled to a `Path`. As such, we have to rewrite `lint_fix`, `lint_only`, etc. to accept some sort of `CodeBlock` struct. Additionally, we have to figure out how to aggregate the diagnostics and display them/their fixes. If we *don't* want to treat them as discrete units, the code will be much simpler (almost akin to the Jupyter implementation).
2. The `CodeBlock` structure, at minimum, should contain the block's content and its offset from the top of the file (what I'm calling its "global offset"). Each diagnostic created while linting the block should have the code block's global offset added to it, meaning that when we display the line to users, they see the actual Markdown line number.
3. I didn't look into this, but we *may* need to expand `UniversalNewLines` to take Markdown line breaks as well, though again, I'm not sure how that works.
4. Unifying the Jupyter/Markdown parsers as `Sources`. I keep going back and forth on what the trait should entail, or if there should even be one.

Pinging @MichaReiser @charliermarsh just in case you guys have time to weigh in. I think the bare minimum question we need to answer is if we want to treat the code blocks as discrete units - I think we should, but that will require quite a bit of work.

P.S. sorry for the poor issue hygiene with the multiple closed PRs. I've been hacking around and didn't realize they're all on here.

**Comment by MichaReiser** [comment]
That's helpful knowledge that you gained with your prototype. Thanks for working on it.

I want to throw in two more use cases:

* We could also lint the markdown, and not just the python code in markdown documents
* We could lint SQL inside of Python files. This is similar to Markdown in that statements are independent from each other but it becomes necessary to call from the python linting into the SQL linting, inverting the direction

The conclusion I'm coming too is that it's probably worth abstracting over the file types instead where Ruff provides a `lint`, `fix`, `format` etc. functions per file type. I don't know what the exact signature would be. These file type specific functions can then pre-process the input and e.g. call into other linters. For example, the markdown linter can call out to the python linter if it finds a python code block and then post-process the diagnostics, to patch up the line numbers. 

The way this would work for SQL is that the python linter calls the SQL linter if it finds a SQL expression when traversing the AST. This is inspired by Prettier's approach where Prettier detects template-literals that are tagged with `graphql` (this is probably not valid syntax, I don't remember all the details):

```javascript
let result = useQuery(graphql`
	query myQuery() {
	}
`);
```

Rome implements something very close to this design. Each language supports different capabilities (we may support linting SQL but not formatting). How these capabilities are implemented is transparent to the CLI. All the CLI cares about is that there's a `lint` function to call and it knows how to lint the files content.

https://github.com/rome/tools/blob/38104b339da8ff688f469799e1fc3f4a46f3d2ec/crates/rome_service/src/file_handlers/javascript.rs#L86-L124

**Comment by paddyroddy** [comment]
@evanrittenhouse I was definitely considering them as discrete units when I raised this. For example here https://github.com/astro-informatics/sleplet/blob/c510795e7ebd91000b8afe53276e522b75cdfbb6/.github/workflows/examples.yml#L33 I use [pytest-codeblocks](https://github.com/nschloe/pytest-codeblocks) to run some example code blocks (essentially acting as another type of test).

**Comment by evanrittenhouse** [comment]
@MichaReiser I agree with the file-agnostic hot path that calls into adapters for different languages. I think the approach you propose sort of mimics LSPs in that each language can have different capabilities, implemented differently. I'd love to work on this, but I think that there will be a lot of design decisions that you guys probably want to weigh in on. Is it even worth continuing work on this before we can have those talks?

**Comment by MichaReiser** [comment]
> I think the approach you propose sort of mimics LSPs in that each language can have different capabilities, implemented differently. 

That's a neat comparison! 

> Is it even worth continuing work on this before we can have those talks?

I'm not sure. What I outlined above is also only what **I** considered doing. There isn't any alignment on the team. Let's maybe first finish the fix refactor. That would allow me to focus on one side project only.

**Comment by mjkanji** [comment]
Chiming in here to ask if the development of this feature can also take Quarto (`.qmd`) into consideration. It's a great project aimed at providing a plaintext equivalent of Jupyter notebooks. 

Hopefully, adding support shouldn't be too complicated because, like (vanilla) Markdown, `.qmd` files denote Python blocks explicitly, albeit with a slightly different syntax:

````md
# This is a Markdown heading

This is regular text.

```{python}
#| echo: false

# This is a Python comment
print("Hello, Ruff!")
```
````

I don't imagine Ruff necessarily cares about specially formatted comments (though maybe it'll become relevant if you add formatting as a core functionality, #1904), but the `#| echo: false` at the top is a Quarto-specific comment that controls its behaviour. See [here](https://quarto.org/docs/computations/execution-options.html#output-options).

Unlike Markdown, Quarto documents are equivalent to Jupyter notebooks so you can (hopefully?) re-use the work being done to add Jupyter support.

**Comment by evanrittenhouse** [comment]
FWIW from my prototypes the Markdown parser will involve large changes to the hot path, so it's a ways down the road. This is because each code block creates a "context" independent of every other code block (vs. a normal Python/Jupyter file where the context is carried throughout the file). The current backend is tied to the assumption that one file contains one context.   

I can't speak to Quarto/Jupyter, though. May be worth raising a separate feature request, since it *sounds* like the work would involve changing some magic strings. 
cc @dhruvmanila (who implemented Jupyter support and is on the core team)

**Comment by RonnyPfannschmidt** [comment]
for sanity it might make absolute sense to consider each markdown fragment a "own file" with a offset starting point

this capability might actually make it easier to make IDE/language server helpers, where one might want to reformat a function/class alone while editing a file
(and the reformatting would basically work off a initial line/column/Position offset, plus the plain text of the lines intended for reformatting)

(im coming from https://github.com/entangled/entangled.py/issues/5)

**Comment by konstin** [comment]
Formatting markdown code blocks is now supported: https://github.com/astral-sh/ruff/pull/9030

**Comment by ddelange** [comment]
tnx for the ping :+1: 
```toml
[tool.ruff.format]
docstring-code-format = true
```

**Comment by paddyroddy** [comment]
> Formatting markdown code blocks is supported in [v0.1.8](https://github.com/astral-sh/ruff/releases/tag/v0.1.8): #9030

Noice! Thank you

**Comment by paddyroddy** [comment]
> Formatting markdown code blocks is supported in [v0.1.8](https://github.com/astral-sh/ruff/releases/tag/v0.1.8): #9030

I believe that it is in the docstrings of functions as opposed to true code blocks in markdown that this issue was about

**Comment by tvatter** [comment]
Just saw this issue and noticed that it would also close #6140 and #8800.

**Comment by Avasam** [comment]
> ### Embedded Languages
> The idea is that a document written in one language can contain code written in another language. Examples are:
> - Code blocks in markdown documents
> - SQL in Python code
> - Python code in Jupyter Notebooks

I'd like to mention HTML as well. Although that would require configuring a selector to find the node element responsible for containing Python code. That's because there is no established standards. The `lang` attribute exists, but it is somewhat out of spec to use it for non human languages (or more specifically, non-valid BCP 47 language tag). Most common solution is to use a custom class for styling anyway, ie `<div class="language-python">`, or a custom element entirely ie `<myapp-code-viewer language="python">`.

Anyway I don't think that should come before Markdown is completely figured out. I just wanted to weigh in on the aspect of supporting Python code blocks in other languages in a generic way.

**Comment by adamtheturtle** [comment]
I have created a tool which enables this use case: [`doccmd`](https://doccmd.readthedocs.io/en/latest/).

For example, you might run:

```sh
$ doccmd --language=python --no-pad-file --command="ruff format" README.md CHANGELOG.rst
$ doccmd --language=python --command="ruff check" README.md CHANGELOG.rst
```

It is new and I'm very open to feedback on it.

By default, `doccmd` basically adds newlines to the beginning of the extracted code block to make line numbers in errors match the documentation file. This doesn't work for `ruff format` which expects to not have a bunch of newlines at the start, so there is the `--no-pad-file` option.

I'm also using it to run `mypy`, `pyright`, `vulture`, `interrogate` ... on my documentation.

**Comment by 34j** [comment]
> I have created a tool which enables this use case: [`doccmd`](https://adamtheturtle.github.io/doccmd/).
> 
> For example, you might run:
> 
> $ doccmd --language=python --no-pad-file --command="ruff format" README.md CHANGELOG.rst
> $ doccmd --language=python --command="ruff check" README.md CHANGELOG.rst
> 
> It is new and I'm very open to feedback on it.
> 
> By default, `doccmd` basically adds newlines to the beginning of the extracted code block to make line numbers in errors match the documentation file. This doesn't work for `ruff format` which expects to not have a bunch of newlines at the start, so there is the `--no-pad-file` option.
> 
> I'm also using it to run `mypy`, `pyright`, `vulture`, `interrogate` ... on my documentation.

pre-commit version:

```yaml
  - repo: https://github.com/adamtheturtle/doccmd-pre-commit
    rev: v2025.4.8
    hooks:
      - id: doccmd
        args:
          ["--language", "python", "--no-pad-file", "--command", "ruff format"]
        additional_dependencies: ["ruff"]
```

**Comment by RonnyPfannschmidt** [comment]
@34j thats a lovely starting point, unfortunately currently pre-commit is not setup in a way where different hooks are in sync about versions, so theres a minor risk of behaviour getting out of sync

on the upside the area of application is different so we shouldn’t see outright conflicts, but there is a chance of inconsistent behaviour

**Comment by adamtheturtle** [comment]
@RonnyPfannschmidt I have a setup like this:

```yaml
      - id: ruff-format-fix-docs
        name: Ruff format docs
        entry: uv run --extra=dev doccmd --language=python --no-pad-file --command="ruff format"
        language: python
        types_or: [markdown, rst]
        additional_dependencies: [uv==0.7.20]
```

This runs locally and on GitHub Actions but not in https://pre-commit.ci, but it does get `ruff` and `doccmd` versions as specified in `pyproject.toml`.

**Comment by RonnyPfannschmidt** [comment]
thats a interesting workaround, but practically sidesteps pre-commit so much that its only used as a jumpstart for an uv command

**Comment by adamtheturtle** [comment]
> thats a interesting workaround, but practically sidesteps pre-commit so much that its only used as a jumpstart for an uv command

It isn't ideal but you do still get a few of the pre-commit features / benefits. For example, it only runs on changed files of the relevant type.

**Comment by 34j** [comment]
> [@34j](https://github.com/34j) thats a lovely starting point, unfortunately currently pre-commit is not setup in a way where different hooks are in sync about versions, so theres a minor risk of behaviour getting out of sync
> 
> on the upside the area of application is different so we shouldn’t see outright conflicts, but there is a chance of inconsistent behaviour

So, are you concerned about a situation where the versions of pre-commit-ruff and the ruff in the doccmd-pre-commit are different? In this case, I think there is a possibility to avoid this by specifying doccmd as additional_dependencies in pre-commit-ruff instead and calling doccmd in entry, but I don't know if it works and don't have the energy to try that.

**Comment by 34j** [comment]
I've tried this and it seems to work:

```yaml
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.12.3
    hooks:
      - id: ruff-format
      - id: ruff-format
        types_or: [markdown, rst]
        require_serial: true
        entry: doccmd --language python --no-pad-file --command "ruff format"
        additional_dependencies: ["doccmd"]
```

**Comment by nijel** [comment]
You can also pin `additional_dependencies` and update them in sync with ruff (https://github.com/WeblateOrg/weblate/commit/fe89a5f7dbd5a74f98dda55d8b64dc5e6058c812 introduces that for Weblate). For example, we automate this using Renovate as configured in https://github.com/WeblateOrg/meta/blob/18e32c7e7dd6af0882b418eb67d495ebd261379f/renovate.json#L80-L90 (which might not be 100% reusable as it assumes that anything with `==` in `.pre-commit-config.yaml` is a Python dependency).

**Comment by MichaReiser** [comment]
Thanks @34j for writing the tool and letting users know about it. I suggest that discussions related to said tool are moved to its repository. This issue is about adding such functionality to ruff and many users are subscribed to it. Thank you

**Comment by DetachHead** [comment]
as a workaround to format python codeblocks in markdown files, you can use the [ruff dprint plugin](https://dprint.dev/plugins/ruff/):

```
uv add --dev dprint-py
dprint config add ruff
dprint config add markdown
```

**Comment by amyreese** [comment]
Hey folks, I'm working on a prototype of this functionality in #22470 and we were discussing potential options for enabling this feature to users. To start, we will be gating the functionality behind the use of `--preview` as its still in an experimental state, and requiring users to manually run the formatter against a set of `.md` files.

Once we have more of the feature set completed and want to stabilize support, there are a few paths we could take:

- Require users who want markdown formatting on-by-default to set `extend-include = ["**/*.md"]` in their project config (opt-in).
- Add a new `format.markdown-code-blocks = true` config option, similar to docstring formatting, that would automatically include `.md` files in the global default search path (opt-in). What do do if the user explicitly passes a `.md` file while the feature is disabled is tbd.
- Turn it on by default, including `.md` in the global default search path, and allow users to set `exclude = ["**/*.md"]` or similar in their project config if they don't want the feature (opt-out).

If you have opinions or preferences, please share them in #22636.

**Comment by DetachHead** [comment]
> Turn it on by default, including `.md` in the global default search path, and allow users to set `exclude = ["**/*.md"]` or similar in their project config if they don't want the feature (opt-out).

i think this is the better choice. i prefer when linters enable all checks by default because it's much better for discoverability. you can be confident that all users who disabled it have done so because they don't want it, and that all users who keep it enabled do.

if this feature was disabled by default, a vast majority of people who don't check changelogs would be missing out on this feature that they would enable if they knew it existed.

----

> If you have opinions or preferences, please share them in https://github.com/astral-sh/ruff/issues/22636.

EDIT: oops i didn't see this. will repost my comment there as well

**Comment by mkniewallner** [comment]
I've played a bit with the feature. First of all, I love it, thanks for working on it!

I'm personally not a huge fan of formatting code blocks without an explicit language, even if they are valid Python.

Running it at work today, we had something like:
````
```
directory/file.py
other_directory/foo/file.txt
```
````

that got reformatted into:
````
```
directory / file.py
other_directory / foo / file.txt
```
````

It is valid Python for the parser (even if at runtime it would not since the variables do not exist), but then a lot of things can be parsed as Python, even if it was not the intention behind the block.

We could argue that `text` should have been added in the example above, but I believe it's rather common to have code blocks without language in markdown files, especially for raw text, so I fear that Ruff would format way more things than it should by default.

I personally find it less problematic that Ruff skips formatting a Python block because someone forgot to add `python` language, rather than it formatting things it should not because Ruff does assumptions about whether it is Python.

Would you consider making that behaviour opt-in rather than by default? Or at the very least, provide an option to disable the formatting of blocks without a language?

**Comment by ntBre** [comment]
Thank you for the feedback! I'm curious what @amyreese thinks, but it sounds reasonable to me to skip blocks without a language specified.

**Comment by ddelange** [comment]
plus one on only formatting `py` and `python` language code blocks (and potentially `pycon` although that is not technically valid python):
```pycon
>>> def fn():
...     return 1
... fn()
1
```

**Comment by amyreese** [comment]
I looked at what `blacken-docs` does in these cases, and it does skip formatting unlabeled blocks. This was originally requested by Micha I think, but I agree that can result in unwanted/undesirable changes.

Formatting of `pycon` blocks is on the eventual wishlist for this feature, but other pieces are higher priority. I opened an issue to help track that.

**Comment by Avasam** [comment]
I agree w/ @mkniewallner, I have a hard time imagining a scenario where Ruff not formatting a code block couldn't be fixed by adding a missing `python` language marker. So I'd rather that be the behaviour.
Imagine if all formatters decided unannotated markdown blocks should be formatted in their language !

Especially for a large project where a formatter is added. Or an existing project where this feature is turned on. It's less friction if unanotated code blocks aren't formatted.

And that's coming from someone vouching for the python-in-markdown formatting to eventually be turned on by default, *and* has a markdown linter to to annotate even "text" code block.

**Comment by harupy** [comment]
`.mdx` support would be nice :)

Original request: https://github.com/astral-sh/ruff/pull/22470#discussion_r2712258383

Can I contribute this?

**Comment by upekkha** [comment]
> `.mdx` support would be nice :)

Please also consider supporting `.markdown` as file extension.

**Comment by kdeldycke** [comment]
I don't know if these alternative file extensions are just aliases or if they implies some format subtleties, but in the mean time, you can direct Ruff to process these files with:
```toml
[tool.ruff]
extend-include = [ "*.{markdown,mdown,mkdn,mdwn,mkd,md,mdtxt,mdtext,mdx}" ]
```

**Comment by ntBre** [comment]
Unfortunately, I don't think that's quite enough. There's a secondary check on the extension beyond including the file. See https://github.com/astral-sh/ruff/pull/22947/changes/abde074124990ef922be27479ad91febeffe8687, for example. It makes sense to me to include at least `mdx` and `markdown`. I'm not sure how common these others are.

**Comment by Avasam** [comment]
For the curious, MDX is a mix of markdown and jsx: https://mdxjs.com / https://github.com/mdx-js/mdx

Now that Ruff supports python code blocks in markdown, I'll hope to see ReStructuredText (rst) eventually ! also quite popular for Python doc because of sphinx

**Comment by amyreese** [comment]
I opened #23204 to track looking for a better way to handle the myriad file extensions for markdown without needing an exhaustive list of possible matches, potentially duplicated in multiple places if/when we stabilize the feature and discover markdown files by default.

**Comment by ddelange** [comment]
@amyreese thanks for pushing this through! what's the timeline for this to move out of preview?

**Comment by MichaReiser** [comment]
Our current plan is to stabilize this feature with the next minor, which we plan to do in the next few weeks.

**Comment by LeoVasanko** [comment]
Quite an annoying feature, and definitely came as a surprise having all my docs formatted. Docs inherently use different formatting than code files. We want to avoid empty lines in markdown, align end of line comments etc, that we would not do in ordinary code. Yes, I can exclude markdown from it, but it really should be opt in to begin with. Somehow I got it simply by upgrading to ruff 0.16.7.

**Comment by DetachHead** [comment]
opt-in is worse than opt-out imo because it makes a feature that most users would probably want less discoverable (wanting to have different formatting in markdown sounds uncommon). users who don't want this behavior can easily revert all the changes it made and disable it

**Comment by LeoVasanko** [comment]
I agree, if a feature is default-off it is used by practically no-one. That being said, before you make it default-on, it needs to have a very good implementation to not cause disruption. I suppose this is for another issue at this point, but the problem is that *default* formatting for documentation and code needs to follow very different rules. Requiring every user to provide those rules separately based on path or file extensions is bad design. But then, having different built-in defaults per format is of course also a complication.

Just a single example of the different expectations. Two spaces before comment rather than trying to align them is definitely the way to go in actual code because this is much easier to maintain through edits. However, in documentation it should be obvious that we want (actual example)

```python
r.pretty    # "Facebook"
r.kind      # "social"
r.provider  # "Meta"
r.url       # "http://www.facebook.com/externalhit_uatext.php"
```

Instead of (as now formatted by ruff)

```python
r.pretty  # "Facebook"
r.kind  # "social"
r.provider  # "Meta"
r.url  # "http://www.facebook.com/externalhit_uatext.php"
```

My hands are full with other projects and I won't even make this an issue, but the problem at least should be in your attention and I do hope that someone intends to work on it for a true solution. No need to reply.
